tss-esapi/src/utils/mod.rs

```diff
@@ -23,6 +23,17 @@ use crate::{Context, Error, Result, WrapperErrorKind};
 use std::convert::TryFrom;
 use zeroize::Zeroize;
 
+#[cfg(all(
+    any(feature = "p256", feature = "rsa",),
+    any(feature = "sha1", feature = "sha2",)
+))]
+mod quote;
+#[cfg(all(
+    any(feature = "p256", feature = "rsa",),
+    any(feature = "sha1", feature = "sha2",)
+))]
```

Comment:
```suggestion
mod quote;
```

Those cfg are not required anymore

Reply:
The standard configuration has no sha1 and sha2 feature enabled.
Without both, the compiler complains about the match, which does not match anything but the wildcard.
Without rsa and without elliptic curve, the compiler also has some complaints.

Reply:
I think I would rather take a couple `#[allow(unused)]` over those combination of cfg, but this is a matter of taste I guess.

<details>

```
diff --git a/tss-esapi/Cargo.toml b/tss-esapi/Cargo.toml
index 05375b055a..6db91d09c3 100644
--- a/tss-esapi/Cargo.toml
+++ b/tss-esapi/Cargo.toml
@@ -49,7 +49,7 @@
 sha3 = { version = "0.10.8", optional = true }
 sm2 = { version = "0.13.3", optional = true }
 sm3 = { version = "0.4.2", optional = true }
-digest = { version = "0.10.7", optional = true }
+digest = { version = "0.10.7", optional = true, features = ["alloc"] }
 signature = { version = "2.2.0", features = ["std"], optional = true}
 cfg-if = "1.0.0"
 strum = { version = "0.26.3", optional = true }
diff --git a/tss-esapi/src/utils/mod.rs b/tss-esapi/src/utils/mod.rs
index 6077182a6e..d4df9c5cae 100644
--- a/tss-esapi/src/utils/mod.rs
+++ b/tss-esapi/src/utils/mod.rs
@@ -23,15 +23,7 @@
 use std::convert::TryFrom;
 use zeroize::Zeroize;

-#[cfg(all(
-    any(feature = "p224", feature = "p256", feature = "p384", feature = "rsa"),
-    any(feature = "sha1", feature = "sha2",)
-))]
 mod quote;
-#[cfg(all(
-    any(feature = "p224", feature = "p256", feature = "p384", feature = "rsa"),
-    any(feature = "sha1", feature = "sha2",)
-))]
 pub use quote::checkquote;

 /// Create the [Public] structure for a restricted decryption key.
diff --git a/tss-esapi/src/utils/quote.rs b/tss-esapi/src/utils/quote.rs
index 4063b5ede3..388c3aa2b0 100644
--- a/tss-esapi/src/utils/quote.rs
+++ b/tss-esapi/src/utils/quote.rs
@@ -8,23 +8,22 @@
     structures::{Attest, AttestInfo, DigestList, PcrSelectionList, Public, QuoteInfo, Signature},
     traits::Marshall,
 };
-use digest::{Digest, DynDigest};
+#[allow(unused)]
+use digest::Digest;
+use digest::DynDigest;

-#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
 use crate::{abstraction::public::AssociatedTpmCurve, structures::EccSignature};
-#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
 use ecdsa::{
     hazmat::{DigestPrimitive, VerifyPrimitive},
     PrimeCurve, SignatureSize, VerifyingKey,
 };
-#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
 use elliptic_curve::{
     generic_array::ArrayLength,
     point::AffinePoint,
     sec1::{FromEncodedPoint, ModulusSize, ToEncodedPoint},
     CurveArithmetic, FieldBytesSize,
 };
-#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
+#[allow(unused)]
 use signature::hazmat::PrehashVerifier;

 #[cfg(feature = "rsa")]
@@ -32,7 +31,7 @@
 #[cfg(feature = "rsa")]
 use signature::Verifier;

-#[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
+#[allow(unused)]
 fn verify_ecdsa<C>(
     public: &Public,
     message: &[u8],
@@ -147,6 +146,7 @@
     }
 }

+#[allow(unused)]
 fn checkquote_pcr_digests(
     quote: &QuoteInfo,
     selections: &PcrSelectionList,
@@ -185,8 +185,8 @@
     if digest_pos != digests.len() {
         return Err(Error::WrapperError(WrapperErrorKind::WrongParamSize));
     }
-    let digest = hasher.finalize().to_vec();
-    if digest != quote.pcr_digest().as_bytes() {
+    let digest = hasher.finalize();
+    if digest.as_ref() != quote.pcr_digest().as_bytes() {
         return Ok(false);
     }
     Ok(true)
@@ -312,10 +312,11 @@
         }
     };

+    #[allow(unused)]
     let bytes = attest.marshall()?;

+    #[allow(unused)]
     let hash_alg = match (public, signature) {
-        #[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]
         (Public::Ecc { parameters, .. }, _) => {
             let mut hash_alg = None;
             macro_rules! impl_check_ecdsa {
```

</details>

Reply:
I totally understand that it looks ugly and I am totally up to a cleaner solution.
Allowing dead code and unused variables does not get rid of all compilation issues. E.g. if sha1 and sha2 are not enabled, the errors include unreachable statements:

```
    let mut hasher: Box<dyn DynDigest> = match hash_alg {
        #[cfg(feature = "sha1")]
        HashingAlgorithm::Sha1 => Box::new(sha1::Sha1::new()),
        #[cfg(feature = "sha2")]
        HashingAlgorithm::Sha256 => Box::new(sha2::Sha256::new()),
        #[cfg(feature = "sha2")]
        HashingAlgorithm::Sha384 => Box::new(sha2::Sha384::new()),
        #[cfg(feature = "sha2")]
        HashingAlgorithm::Sha512 => Box::new(sha2::Sha512::new()),
        _ => {
            return Err(Error::WrapperError(WrapperErrorKind::UnsupportedParam));
        }   
    };  
```

Results in
```
error: unreachable statement                                                                                                                                               
   --> tss-esapi/src/utils/quote.rs:175:5                                                                                                                                  
    |                                                                                                                                                                      
161 |       let mut hasher: Box<dyn DynDigest> = match hash_alg {                                                                                                          
    |  __________________________________________-                                                                                                                         
162 | |         #[cfg(feature = "sha1")]                                                                                                                                   
163 | |         HashingAlgorithm::Sha1 => Box::new(sha1::Sha1::new()),                                                                                                     
164 | |         #[cfg(feature = "sha2")]                                                                                                                                   
...   |                                                                                                                                                                    
173 | |     };                                                                                                                                                             
    | |_____- any code following this `match` expression is unreachable, as all arms diverge                                                                               
174 |                                                                                                                                                                      
175 | /     for selection in selections.get_selections() {                                                                                                                 
176 | |         let sel_count = selection.selected().len();                                                                                                                
177 | |         if digest_pos + sel_count > digests.len() {                                                                                                                
178 | |             return Err(Error::WrapperError(WrapperErrorKind::WrongParamSiz...                                                                                      
...   |                                                                                                                                                                    
184 | |     }                                                                                                                                                              
    | |_____^ unreachable statement
```

Reply:
Also if the result of this discussion is to remove (some) ifdef and allow dead code and unused variables, ci-tests should be excluded, so we still catch unclean code. (I still think the current version is cleaner in total)

Maybe we should also add test compilation with different features enabled during ci-test.

Reply:
This unreachable statement we can probably sort out by moving this to a function, but I get what you're saying.

Reply:
https://github.com/brandsimon/rust-tss-esapi/pull/8

Reply:
All the features should be tested automatically. I made an effort a while ago to make it happen by having the crate itself as `dev-dependency` with all features we enabled. I forgot to add some documentation though to specify that all new options needs to be added to the list.

Reply:
@baloo 
Your branch fails to compile with default features.
```
error: unreachable statement                                                                                                                                               
   --> tss-esapi/src/utils/quote.rs:398:5                                                                                                                                  
    |                                                                                                                                                                      
334 |       let hash_alg = match (public, signature) {                                                                                                                     
    |  ____________________-                                                                                                                                               
335 | |         #[cfg(any(feature = "p224", feature = "p256", feature = "p384"))]                                                                                          
336 | |         (Public::Ecc { parameters, .. }, _) => {                                                                                                                   
337 | |             let mut hash_alg = None;                                                                                                                               
...   |                                                                                                                                                                    
396 | |     };                                                                                                                                                             
    | |_____- any code following this `match` expression is unreachable, as all arms diverge                                                                               
397 |                                                                                                                                                                      
398 | /     if qualifying_data != attest.extra_data().as_bytes() {                                                                                                         
399 | |         return Ok(false);                                                                                                                                          
400 | |     }                                                                                                                                                              
    | |_____^ unreachable statement
```

Reply:
If no one has hard feelings about this, I'd prefer to keep the cfgs for now.

Reply:
@Superhepper Do you mean all features enabled or for every feature combination? (I meant all combinations, I created a separate issue for this